vega_sim/network_service.py

```diff
@@ -337,6 +342,83 @@ def core_client(self) -> None:
         )
         pass
 
+    def check_datanode(self, raise_on_error: Optional[bool] = True):
+        """Checks if the current data-node connection is healthy.
+
+        If the current data-node connection has timed out or the end-point is
+        unresponsive then the service will attempt to establish a connection
+        with a new healthy data-node.
+
+        Args:
+            max_attempts (int, optional):
+                Maximum number of connection attempts to attempt before raising an
+                error. Defaults to -1 (infinite attempts).
+        """
+
+        try:
+            self.ping_datanode()
+            return
+
+        except grpc.FutureTimeoutError:
+            # Unable to connect to datanode
+            msg = f"Connection to endpoint {self._data_node_grpc_url} timed out."
+            if raise_on_error:
+                raise RuntimeError(msg)
+            else:
+                logging.warning(msg)
+                self.switch_datanode()
+
+        except grpc._channel._InactiveRpcError:
+            msg = f"Connection to endpoint {self._data_node_grpc_url} in active."
+            if raise_on_error:
+                raise RuntimeError(msg)
+            else:
+                logging.warning(msg)
+                self.switch_datanode()
+
+    def switch_datanode(self, max_attempts: Optional[int] = -1):
+        """Attempts to establish a new data-node connection.
+
+        Args:
+            max_attempts (int, optional):
+                Maximum number of connection attempts to attempt before raising an
+                error. Defaults to -1 (infinite attempts).
+        """
+
+        attempts = 0
+        while attempts != max_attempts:
+            attempts += 1
+
+            try:
+                self.datanode_index += 1
+                if self.datanode_index >= len(
+                    self.network_config["API"]["GRPC"]["Hosts"]
+                ):
+                    self.datanode_index = 0
```

Comment:
Let's shorten this to `datanode_index = (datanode_index + 1) % len(self.network_config["API"]["GRPC"]["Hosts"])`

Reply:
Actually, let's use `itertools.cycle(self.network_config["API"]["GRPC"]["Hosts"])` then you can just call `next` on it until the number of tries has been reached